src/main/java/au/org/ala/biocache/dao/SearchDAOImpl.java

```diff
@@ -911,120 +1054,118 @@ public void writeCoordinatesToStream(SearchRequestParams searchParams,OutputStre
-                                        }
+                                        // throttle the download by sleeping
+                                        Thread.sleep(localThrottle);
                                     }
                                     qr = runSolrQuery(splitByFacetQuery, null, downloadBatchSize, startIndex, "_docid_", "asc");
                                 } else {
                                     qr = null;
                                 }
                             }
-                            return recordsForThread;
+                            return recordsForThread.get();
                         }
                     };
-                    futures.add(pool.submit(solrCallable));
+                    solrCallables.add(solrCallable);
                 }
     
-                //check the futures until all have finished
-                int totalDownload = 0;
-                Set<Future<Integer>> completeFutures = new HashSet<Future<Integer>>();
-                boolean allComplete = false;
-                while(!allComplete){
-                    for(Future future: futures){
-                        if(!completeFutures.contains(future)){
-                            if(future.isDone()){
-                                totalDownload += (Integer) future.get();
-                                completeFutures.add(future);
-                            }
-                        }
-                    }
-                    allComplete = completeFutures.size() == futures.size();
-                    if(!allComplete){
-                        Thread.sleep(1000);
+                // TODO: Do we support downloads running for longer than 1 day??
+                List<Future<Integer>> futures = nextExecutor.invokeAll(solrCallables, 1, TimeUnit.DAYS);
+                AtomicInteger totalDownload = new AtomicInteger(0);
+                for(Future<Integer> future: futures){
+                    if(future.isDone()){
+                        totalDownload.addAndGet(future.get());
                     }
                 }
-                pool.shutdown();
-                out.flush();
 
                 long finish = System.currentTimeMillis();
                 long timeTakenInSecs = (finish-start)/1000;
                 if(timeTakenInSecs <= 0) timeTakenInSecs = 1;
-                logger.info("Download of " + resultsCount + " records in " + timeTakenInSecs + " seconds. Record/sec: " + resultsCount.intValue()/timeTakenInSecs);
+                if(logger.isInfoEnabled()) {
+                    logger.info("Download of " + resultsCount + " records in " + timeTakenInSecs + " seconds. Record/sec: " + resultsCount.intValue()/timeTakenInSecs);
+                }
             } finally {
-                rw.finalise();
+                try {
+                    concurrentWrapper.finalise();
+                } finally {
+                    try {
+                        writerThread.join(100000);
```

Comment:
A new ArrayBlockingQueue is setup for each writeResultsFromIndexToStream method call, so there is no contention across queries from different users at that level. 

The Callable object created to fetch each Facet result however will have contention for the threads in the shared ExecutorService fixed thread pool, which has its number of threads setup using configuration parameter solr.downloadquery.maxthreads (defaults to 30). This is intentional to prevent it spiking to 400+ as it does right now, which causes the system to become unresponsive to all operations.

I will look into timing out on the supply side using BlockingQueue.offer(e, long, TimeUnit) instead of BlockingQueue.put . 

Timing out within the consumer thread isn't possible currently with the RecordWriter API, and would require yet another monitor thread (and a variable) to busy-wait check on a consumer's last written timestamp which would then interrupt the consumer thread if it was idle for a certain amount of time. That sounds very complicated and would interrupt other threads to do its work. We should be able to timeout using thread interrupts from the producer threads to the consumer thread to avoid that.